StRoot/StFcsDbMaker/StFcsDb.cxx

```diff
@@ -1979,3 +1979,125 @@ void StFcsDb::readGainCorrFromText(){
   }
   fclose(F);
 }
+
+//g2t track info                                                                                                   
+unsigned int StFcsDb::backTraceG2tTrack(unsigned int id, g2t_track_st* g2ttrk){
+  int i = id - 1;
+  while(g2ttrk[i].next_parent_p !=0){
+    if(mDebug>3)
+      LOG_INFO<<Form("  BackTrace from=%3d id=%3d Epid=%4d Gpid=%3d Vtx=%3d Parent=%3d E=%6.2f",
+                     id,g2ttrk[i].id,g2ttrk[i].eg_pid,g2ttrk[i].ge_pid,g2ttrk[i].start_vertex_p,
+		     g2ttrk[i].next_parent_p,g2ttrk[i].e)<<endm;
+    i = g2ttrk[i].next_parent_p - 1;
+  }
+  if(mDebug>3) LOG_INFO<<Form("  BackTrace from=%3d id=%3d Epid=%4d Gpid=%3d Vtx=%3d Parent=%3d E=%6.2f Primary!!",
+			      id,g2ttrk[i].id,g2ttrk[i].eg_pid,g2ttrk[i].ge_pid,g2ttrk[i].start_vertex_p,
+			      g2ttrk[i].next_parent_p,g2ttrk[i].e)<<endm;
+  return i + 1;
+}
+    
+const g2t_track_st* StFcsDb::getParentG2tTrack(StFcsHit* h, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
+  const vector<pair<unsigned int, float>>& gt = h->getGeantTracks();  
+  ntrk=gt.size();
+  float detot=0;
+  if(order > gt.size()) {fraction=0; return 0;}
+  for(int itrk=0; itrk<ntrk; itrk++) detot += gt[itrk].second;
+  fraction = gt[order].second / detot;
+  if(mDebug>3){
+    for(unsigned int jtrk=0; jtrk<ntrk; jtrk++){
+      LOG_INFO << Form("Hit's parent G2T Track %3d id=%3d dE=%f",jtrk,gt[jtrk].first,gt[jtrk].second)<<endm;
+    }
+  }
+  return &g2ttrk[gt[order].first-1];
+}
+
+const g2t_track_st* StFcsDb::getPrimaryG2tTrack(StFcsHit* h, g2t_track_st* g2ttrk, float& fraction, int& ntrk,unsigned int order){
+  const vector<pair<unsigned int, float>>& gt = h->getGeantTracks();
+  vector<pair<unsigned int, float>> primary;
+  int n=gt.size();
+  float detot=0;
+  for(int itrk=0; itrk<n; itrk++){
+    float de=gt[itrk].second;
+    unsigned int id=backTraceG2tTrack(gt[itrk].first, g2ttrk);
+    int found=0;
+    unsigned int np=primary.size();
+    for(unsigned int jtrk=0; jtrk<np; jtrk++){
+      if(primary[jtrk].first == id) {primary[jtrk].second += de; found=1; break;}
+    }
+    if(found==0) primary.push_back(make_pair(id,de));
```

Comment:
If I understand correctly what you are trying to accomplish here, the logic can be replaced by a single line of code such as:

```c++
primary[id] += de;
```

if `primary` is defined as `unordered_map<unsigned int, float> primary;`. The access to the elements of an `unordered_map` is very efficient O(1)

Reply:
Yes, I don't understand @klendathu2k's answer above. This does not participate in the simulation data flow. The rest is just a premature optimization.

Reply:
@veprbl not sure which answer are you referring to?  (this discussion thread is getting a bit long).

Reply:
It was your original response to suggestion to use a map type from STL:
https://github.com/star-bnl/star-sw/pull/379#discussion_r933201592
> For the most part, the hits from a single track which initiates an EM or hadronic cascade will be presented to the slow simulator before the next track is processed. So the vector is the natural choice here. For any geant hit added to the FCS hit, the pair which needs to be updated is likely to be at the back of the vector. Once all hits have been processed, we can loop over the FCS hits and ensure that they are sorted descending in energy.

After reading this, at least I left with impression that the `vector` type is somehow justified by downstream code or somehow more performant. Hence my suggestion to write code as simple and short as possible so that people can understand it before we optimize anything.

Reply:
One of your impressions is correct.  The vector implementation will be more performant than the other options.  (Or at least I expect it to be given the order in which GEANT creates, and starsim stores, the hits).  

Storing the ID truth pair (ID track, dE contribution) in order of the highest contribution to the hit was chosen by Akio, and I agree that it is the most natural ordering.  This is what drives the choice of a vector rather than a set or a map.

Reply:
I'm not sure if it's worth discussing performance of the code that casually calls `std::sort`.

Reply:
Please suggest alternative